lib/fizzy/utf8.cpp

```diff
@@ -0,0 +1,156 @@
+#include "utf8.hpp"
+#include <cassert>
+
+/*
+ * The Unicode Standard, Version 6.0
+ * (https://www.unicode.org/versions/Unicode6.0.0/ch03.pdf)
+ *
+ * Page 94, Table 3-7. Well-Formed UTF-8 Byte Sequences
+ *
+ * +--------------------+------------+-------------+------------+-------------+
+ * | Code Points        | First Byte | Second Byte | Third Byte | Fourth Byte |
+ * +--------------------+------------+-------------+------------+-------------+
+ * | U+0000..U+007F     | 00..7F     |             |            |             |
+ * +--------------------+------------+-------------+------------+-------------+
+ * | U+0080..U+07FF     | C2..DF     | 80..BF      |            |             |
+ * +--------------------+------------+-------------+------------+-------------+
+ * | U+0800..U+0FFF     | E0         | A0..BF      | 80..BF     |             |
+ * +--------------------+------------+-------------+------------+-------------+
+ * | U+1000..U+CFFF     | E1..EC     | 80..BF      | 80..BF     |             |
+ * +--------------------+------------+-------------+------------+-------------+
+ * | U+D000..U+D7FF     | ED         | 80..9F      | 80..BF     |             |
+ * +--------------------+------------+-------------+------------+-------------+
+ * | U+E000..U+FFFF     | EE..EF     | 80..BF      | 80..BF     |             |
+ * +--------------------+------------+-------------+------------+-------------+
+ * | U+10000..U+3FFFF   | F0         | 90..BF      | 80..BF     | 80..BF      |
+ * +--------------------+------------+-------------+------------+-------------+
+ * | U+40000..U+FFFFF   | F1..F3     | 80..BF      | 80..BF     | 80..BF      |
+ * +--------------------+------------+-------------+------------+-------------+
+ * | U+100000..U+10FFFF | F4         | 80..8F      | 80..BF     | 80..BF      |
+ * +--------------------+------------+-------------+------------+-------------+
+ */
+
+// Ranges for 2nd byte
+enum class Rule
+{
+    Range80BF,  // 80..BF
+    RangeA0BF,  // A0..BF
+    Range809F,  // 80..9F
+    Range90BF,  // 90..BF
+    Range808F,  // 80..8F
+};
+
+namespace fizzy
+{
+bool utf8_validate(const uint8_t* pos, const uint8_t* end) noexcept
+{
+    while (pos < end)
+    {
+        unsigned required_bytes = 1;
+        auto rule = Rule::Range80BF;
```

Comment:
just an idea for better name
```suggestion
        auto byte2_rule = Rule::Range80BF;
```